### ISSUE TITLE
Log Monitor TypeError

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#12565]
No reproducible script (never seen this before), but this happened in the middle of my training:

```
2020-12-02 07:26:37,751	WARNING worker.py:1011 -- The log monitor on node ip-172-31-18-179 failed with the following error:
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/log_monitor.py", line 354, in <module>
    log_monitor.run()
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/log_monitor.py", line 275, in run
    self.open_closed_files()
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/log_monitor.py", line 164, in open_closed_files
    self.close_all_files()
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/log_monitor.py", line 102, in close_all_files
    os.kill(file_info.worker_pid, 0)
TypeError: an integer is required (got type str)
```

cc @rkooo567

**Comment by FarzanT** [comment]
Hi, I still got the same error (randomly) on a multi-node cluster setup where I had to turn the dashboard off. After about 7 hours of processing, the error caused the whole job to terminate. I'm using `Ray 2.0.0dev0`

**Comment by FarzanT** [comment]
I'm also getting the following error (comes afterwards), but I'm not sure which one is causing which, if that's the case.
```
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m Traceback (most recent call last):
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m   File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/multiprocessing/resource_sharer.py", line 142, in _serve
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m     with self._listener.accept() as conn:
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m   File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/multiprocessing/connection.py", line 456, in accept
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m     answer_challenge(c, self._authkey)
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m   File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/multiprocessing/connection.py", line 742, in answer_challenge
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m     message = connection.recv_bytes(256)         # reject large message
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m   File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/multiprocessing/connection.py", line 216, in recv_bytes
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m     buf = self._recv_bytes(maxlength)
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m   File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/multiprocessing/connection.py", line 407, in _recv_bytes
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m     buf = self._recv(4)
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m   File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/multiprocessing/connection.py", line 379, in _recv
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m     chunk = read(handle, remaining)
^[[2m^[[36m(pid=20558, ip=172.16.190.50)^[[0m ConnectionResetError: [Errno 104] Connection reset by peer
```

**Comment by rkooo567** [comment]
Hmm interesting.

**Comment by rkooo567** [comment]
Btw, log_monitor is not related to the dashboard (it is always turned on regardless of the dashboard), and the only impact of this error is that your drivers cannot receive logs from workers.

**Comment by rkooo567** [comment]
Anyway this seems to be a bad bug, so we should take a look at it.

**Comment by rkooo567** [comment]
The following error seems to be unrelated... are you using multi processing pool within a ray task or actor?

**Comment by FarzanT** [comment]
Yes, I'm using Pytorch's data.DataLoader() with num_workers=4. Do you think this is the cause for the latter error?

**Comment by rkooo567** [comment]
It is hard to say, but it looks like the error just seems to be application-specific error (rather than system issues). Do you have any log that indicates this node (172.16.190.50) has an issue?

**Comment by rkooo567** [comment]
Maybe this? https://github.com/pytorch/pytorch/issues/1551

**Comment by FarzanT** [comment]
Perhaps, I'll disable multithreading and will update you! Thanks

**Comment by FarzanT** [comment]
Now I don't get the `ConnectionResetError` anymore, but the same error persists:
```
    os.kill(file_info.worker_pid, 0)
TypeError: an integer is required (got type str)
```

**Comment by FarzanT** [comment]
@rkooo567 What `str` values can `worker_pid` have other than `raylet` and `gcs_server`? I think this is the primary cause of the problem: https://github.com/rkooo567/ray/blob/c197f4cb0f024e44aa83a14939d9b87a1c750ac3/python/ray/log_monitor.py#L100
```
if (file_info.worker_pid != "raylet"
        and file_info.worker_pid != "gcs_server"):
    os.kill(file_info.worker_pid, 0)
```

**Comment by rkooo567** [comment]
Oh good discovery. I think we recently added `autoscaler`. cc @wuisawesome Can you confirm?

**Comment by wuisawesome** [comment]
The timing here is a little strange, since it sounds like Richard discovered this in december but #13434 was merged in january. maybe it's even simpler and we're getting the string representation of an int?

**Comment by ericl** [comment]
Can we just add "autoscaler" to fix this? https://github.com/rkooo567/ray/blob/c197f4cb0f024e44aa83a14939d9b87a1c750ac3/python/ray/log_monitor.py#L100

**Comment by wuisawesome** [comment]
Does anyone have a repro? If this was happening in december, then "autoscaler" was probably not the string breaking this.

Alternatively we could just do 

```
try:
    os.kill(int(file_info.worker_pid), 0)
catch ...
```

**Comment by rkooo567** [comment]
Talked to Alex. Last year, the root cause was that I made a mistake while I changed log monitor code. I think the new issue must be the autoscaler string.

**Comment by FarzanT** [comment]
@rkooo567 Great, so the only thing required is to add `"autoscaler"` here?
```
if (file_info.worker_pid != "raylet"
        and file_info.worker_pid != "gcs_server" and file_info.worker_pid != "autoscaler"):
    os.kill(file_info.worker_pid, 0)
```
I can submit a pull request, I would like to practice!

**Comment by rkooo567** [comment]
Thanks for the immidate action @FarzanT!!!

**Comment by FarzanT** [comment]
I think there might be another problem, I went ahead and changed the same line as in the pull request manually in `log_monitor.py` on our cluster, and I still get this error:
```
2021-02-22 20:58:27,272 WARNING worker.py:1107 -- The log monitor on node mist007.scinet.local failed with the following error:
Traceback (most recent call last):
  File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/site-packages/ray/log_monitor.py", line 359, in <module>
    try:
  File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/site-packages/ray/log_monitor.py", line 280, in run
    self.update_log_filenames()
  File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/site-packages/ray/log_monitor.py", line 167, in open_closed_files
    # If we can't open any more files. Close all of the files.
  File "/home/l/lstein/ftaj/.conda/envs/drp1/lib/python3.7/site-packages/ray/log_monitor.py", line 102, in close_all_files
    and file_info.worker_pid != "autoscaler"):
TypeError: an integer is required (got type str)
```

Maybe there's some other `str`? I can only think of adding a print statement and report what I can find.

**Comment by rkooo567** [comment]
Can you try with this change and give us what's the output? https://github.com/ray-project/ray/pull/14271

**Comment by rkooo567** [comment]
Just a followup, @FarzanT did you see any issue so far?

**Comment by FarzanT** [comment]
@rkooo567 Surprisingly not yet, I submit 24-hour jobs, usually they would throw this error after 7-8 hours. So far so good, but I'll update you as soon as I get something!

**Comment by wuisawesome** [comment]
@FarzanT any luck reproducing?

btw is it possible that the run from 2 days ago was on an old cluster? A little suspicious that the failing line of code isn't the `os.kill` anymore

**Comment by FarzanT** [comment]
@wuisawesome No, thankfully didn't have that problem anymore! 
I think the job was already running when I went ahead and changed the line (sorry I had many jobs at that point). I think the computer copies the code at import once and doesn't care if it's changed in the middle of running a job, correct? In this case, the job was using the older version. I think adding the "autoscaler" string has fixed this issue!

**Comment by wuisawesome** [comment]
Awesome! I'll reclose this issue then :)

**Comment by mmchougule** [comment]
@wuisawesome when will this be released? We are seeing this issue now. I see the last release ray 1.2.0 was before the PR got merged.

**Comment by wuisawesome** [comment]
The fix is in master now. It will also be in Ray 1.3.0. The ETA for 1.3 is around 1-2 weeks (the release process currently takes time)

**Comment by mmchougule** [comment]
Okay, I will try to install ray from `master` and see if this issue is fixed.

**Comment by rkooo567** [comment]
The fix must be included in 1.3 release (which will be release in 1~2 weeks)

**Comment by yshvrdhn** [comment]
When does this actually get hit ? If the PIDs go very large ? I am facing teh same issue if I try to run on say 50 actor instances it runs fine but if try to run on say 22000 actor instances it hits this issue.
